fix: flaky results: same start and end marker

M4tteoP opened the pull request:
I'm investigating some flakiness and found that the same marker might be retrieved for both the starter and the final one. When this happens, the function responsible for retrieving all the marked lines can not locate the starter marker. As a result, it uses all the logs up to the beginning of the file and processes IDs from lines belonging to previous tests, leading to incorrect results. 
My idea is that it can happen when go-ftw is too fast and reads again the same line before the engine/connector manages to write/flush the new logs (which include the final marker) into the file.

From a visual perspective (_Note: The read is performed backward._):



**] Expected scenario:**
start marker: `8888-8888`
final marker: `9999-9999`
```
...
log id:10
8888-8888
log id:11
9999-9999
```
1) We look for the final marker
2) Se start collecting log lines
3) We stop when reaching the start marker
4) log id:10 line is not collected, being out of the markers.


**] Assuming the case that both the start marker and final marker have the same value (equal to the start marker):**
start marker: `8888-8888`
final marker: `8888-8888`
```
...
log id:10
8888-8888
log id:11
9999-9999
```
1) We look for the final marker, in that case, the line `8888-8888` matches
2) From there we start collecting logs
3) `log id:11` is not collected, but rather `log id:10` and all the previous ones are collected.

The PR proposes to:
1) Add a debug log when the start marker is not found and unbounded logs (up to the beginning of the file) are used
2) Introduce a retry mechanism that, when looking for the final marker, repeats the seeking of the log file if the final marker found is equal to the start maker.
3) Do not introduce changes so far if the final marker is found equal to the start marker after some attempts, but print a verbose Warn log line.

It would allow us to have more visibility on the issue and bring in some more resiliency. Anyways I think that it is not expected to have the same start and final marker line, therefore I think adding a check would be beneficial.

I will follow with a test/reproducer whenever I manage to reproduce it in an easy and consistent way

theseion commented:
What you want already exists:
1. `readLimit` limits the amount of backwards scanning to avoid reading the entire log file if it is large
2. if the start marker can't be found, another request is sent to try and flush the log and then the search is repeated, up to `maxMarkerRetries`

`CheckLogForMarker` returns the found line in each case and this line is then used as start / end marker. Even if they are identical, the search logic in `getMarkedLines` should ensure that finding the lines between the markers works.

I see one possible issue: when the log is not flushed immediately, `markAndFlush` will produce multiple marker lines in the log. If the start marker line is identical to these (e.g., there's no differing timestamp) then the these lines would match the start marker line and there wouldn't exist any other log lines in between. Example:

```
<marker ID>
log 1
log 2
<marker ID>
<marker ID>
```

We can solve this problem by using a suffix to the UUID (I'd like to keep a single UUID to make it easier to identify the matching markers when reading the log manually). E.g., the start marker could be `<UUID>-s` and the end marker `<UUID>-e`.

M4tteoP commented:
Thanks Max!

> 1. readLimit limits the amount of backwards scanning to avoid reading the entire log file if it is large
 
 In this scenario `readLimit` would just reduce the amount of unbounded log lines read, but, in the first place, we are reading the wrong log lines. So, with `readLimit` we would just read fewer wrong lines.

 > 2. if the start marker can't be found, another request is sent to try and flush the log and then the search is repeated, up to maxMarkerRetries.
 
 This logic does no help in this scenario: the start marker is correctly found, so no retries are performed.
 
 
> CheckLogForMarker returns the found line in each case and this line is then used as start / end marker. Even if they are identical, the search logic in getMarkedLines should ensure that finding the lines between the markers works.
 
If I'm getting the code right, `getMarkedLines` search logic is not working properly with two identical markers. So we should:

- Never give to `getMarkedLines` two identical markers.
- Or adapt `getMarkedLines` to not really look for the identical markers, but just rely on the UUID (but then there is no point on passing to that function the two marker lines because it will have to find them on its own).

 
Going step by step:
1. `markAndFlush` is called, a request is performed to log the starter marker. Now logs will be like this:
```
...
log 0   <-- old log belonging to the previous request
[msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "WWAVDNpnKkXiHleWcdD"]
```
2. `markAndFlush` internally calls `CheckLogForMarker` which looks, starting from the end, for a line with `7d16c799-954f-4338-81c9-fb39b32a9521`. The line is found and identified as the start marker line. No retries are performed.
3. `RunStage` stores the start marker via `SetStartMarker`. The start marker now is `[msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "WWAVDNpnKkXiHleWcdD"]`.
4. `RunStage` performs the real request that might or might not trigger some rules and the related logs will be printed.
5. `markAndFlush` is called again to print the end marker. The request is performed correctly, but the Engine is not fast enough on writing rules logs and the end marker.
6. `markAndFlush` internally calls `CheckLogForMarker` when logs are looking like the following:
```
...
log 0 
[msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "WWAVDNpnKkXiHleWcdD"]
log 1
log 2
```
We can notably see that we don't have yet a trailing marker with `7d16c799-954f-4338-81c9-fb39b32a9521` but a different unique_id.

7. `CheckLogForMarker` looks again for a line with `7d16c799-954f-4338-81c9-fb39b32a9521` and within `readLimit` lines, it finds `[msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "WWAVDNpnKkXiHleWcdD"]`
8. `RunStage` stores the final marker via `SetEndMarker` using the same line of the start marker.
Start marker: `[msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "WWAVDNpnKkXiHleWcdD"]`
End marker: `[msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "WWAVDNpnKkXiHleWcdD"]`

9. At this point `getMarkedLines` is called, providing start and end marker.
10. (not so relevant, this point might not happen and the outcome would be the same) In the meanwhile the Engine might managed to write the whole logs. So they might now look like the following:
```
...
log 0 
[msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "WWAVDNpnKkXiHleWcdD"]
log 1
log 2
log 3
[msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "gfrFHrMGygNvrCkkWRY"]
```
11. `getMarkedLines` starts from the bottom and looks for the end marker. The real end marker (`gfrFHrMGygNvrCkkWRY`) and all the logs related to this request are skipped. It finally finds the end marker.
```
...
log 0 
⬆ [msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "WWAVDNpnKkXiHleWcdD"]
❌ log 1
❌ log 2
❌ log 3
❌ [msg "X-CRS-Test 7d16c799-954f-4338-81c9-fb39b32a9521"][unique_id "gfrFHrMGygNvrCkkWRY"]
```
12. From `WWAVDNpnKkXiHleWcdD` it starts reading old log lines (such as `log 0`) until:
- it finds the start marker: but it will never happend, because the start marker has been misused as end marker
- or it reaches `readLimit` reading `readLimit` wrong log lines
- or it reaches the beginning of the document.
13. All these log lines are then used to extract IDs and computing wrong outcomes. The correct logs to be used (`log 1`, `log 2`, `log 3`) are not taken into account

theseion commented:
Thanks for the analysis. You're right, in that case both markers would be the same line. Both cases, the one you described and the one I came up with, would be fixed by using identifying suffixes for the start and end markers, since `CheckLogForMarker` could no longer confuse the two.

M4tteoP commented:
Thank you! It sounds like a very reasonable idea to have the same `uuid` just followed by a suffix.
I have:
- Implemented the suffix idea
- Removed the SeekAttempts
- Kept some more debug verbosity when markers are not found
- Added some logic to not consider as marked lines all the lines found before finding the end marker (backward read). The tests were adding one or more empty lines, in real-case scenarios I guess there could be logs that we should just ignore

theseion commented on runner/run.go:
```diff
@@ -166,7 +173,8 @@ func RunStage(runContext *TestRunContext, ftwCheck *check.FTWCheck, testCase sch
 
 	if notRunningInCloudMode(ftwCheck) {
-		startMarker, err := markAndFlush(runContext, &testInput, stageId)
+		startUUID := stageId + startUUIDSuffix
```
```suggestion
		startId := stageId + startUUIDSuffix
```

theseion commented on runner/run.go:
```diff
@@ -22,6 +22,13 @@ import (
+const (
+	// start and end UUID suffixes are used to disambiguate start and end markers.
+	// It permits to make the markers unique, while still maintaining one UUID per stage.
+	startUUIDSuffix = "-s"
+	endUUIDSuffix   = "-e"
```
```suggestion
	startUuidSuffix = "-s"
	endUuidSuffix   = "-e"
```

theseion commented on runner/run.go:
```diff
@@ -190,7 +198,8 @@ func RunStage(runContext *TestRunContext, ftwCheck *check.FTWCheck, testCase sch
 
 	if notRunningInCloudMode(ftwCheck) {
-		endMarker, err := markAndFlush(runContext, &testInput, stageId)
+		endUUID := stageId + endUUIDSuffix
```
```suggestion
		endId := stageId + endUUIDSuffix
```

theseion commented on waflog/read.go:
```diff
@@ -158,24 +163,34 @@ func (ll *FTWLogLines) getMarkedLines() [][]byte {
-			endFound = true
+
+		if !endFound {
+			// Skip lines until we find the end marker. Reading backwards, the line we are looking for are
```
```suggestion
			// Skip lines until we find the end marker. Reading backwards, the lines we are looking for are
```

theseion commented on waflog/read.go:
```diff
@@ -158,24 +163,34 @@ func (ll *FTWLogLines) getMarkedLines() [][]byte {
 		ll.markedLines = append(ll.markedLines, saneCopy)
 	}
+	if !startFound {
+		log.Debug().Msgf("starter marker not found while getting marked lines")
```
```suggestion
		log.Debug().Msg("starter marker not found while getting marked lines")
```

theseion commented on waflog/read.go:
```diff
@@ -158,24 +163,34 @@ func (ll *FTWLogLines) getMarkedLines() [][]byte {
 // stageID is the ID of the current stage, which is part of the marker line
+// At this point, stageID includes also the start or end suffix.
 // readLimit is the maximum numbers of lines to check
 func (ll *FTWLogLines) CheckLogForMarker(stageID string, readLimit uint) []byte {
```
```suggestion
func (ll *FTWLogLines) CheckLogForMarker(markerId string, readLimit uint) []byte {
```

theseion commented on waflog/read.go:
```diff
@@ -158,24 +163,34 @@ func (ll *FTWLogLines) getMarkedLines() [][]byte {
 }
 
 // CheckLogForMarker reads the log file and searches for a marker line.
 // stageID is the ID of the current stage, which is part of the marker line
+// At this point, stageID includes also the start or end suffix.
```
```suggestion
// markerId is the ID of the current stage + suffix (for start / end), which is part of the marker line
```

theseion commented on waflog/read_test.go:
```diff
@@ -201,12 +201,15 @@ func (s *readTestSuite) TestFTWLogLines_Contains() {
 	s.NotNil(cfg)
 
 	stageID := "dead-beaf-deadbeef-deadbeef-dead"
```
```suggestion
	stageId := "dead-beaf-deadbeef-deadbeef-dead"
```

theseion commented on waflog/read_test.go:
```diff
@@ -368,12 +373,13 @@ func (s *readTestSuite) TestFindAllIdsInLogs() {
 	s.NotNil(cfg)
 
 	stageID := "dead-beaf-deadbeef-deadbeef-dead"
```
```suggestion
	stageId := "dead-beaf-deadbeef-deadbeef-dead"
```

theseion reviewed (changes requested):
We don't have a test yet, that specifically reproduces the issue, right? Could you add one?

fzipi commented:
@M4tteoP Can you address comments here and fix the conflict so we merge this one?

M4tteoP commented:
Yes, It is in my plans, I will be back on this one soon

M4tteoP commented:
It should finally be ready for another look. Apologies for the delay!

M4tteoP commented on check/base.go:
```diff
@@ -116,6 +116,10 @@ func (c *FTWCheck) GetTriggeredRules() []uint {
 	if c.CloudMode() {
 		return nil
 	}
+	// When a test is expecting to trigger an error and it effectively does, markers are not set.
```
After https://github.com/coreruleset/go-ftw/pull/432, `GetTriggeredRules()` is now called not only when we are looking for IDs in the logs based on the test expectations but also at the end of the test to provide the list of triggered rules. When a test is expected to trigger an error, start and end markers might not populated. As a result, calling TriggeredRules() in such cases triggers a fatal error. Similar to how [we do not support looking for IDs when expecting an error](https://github.com/coreruleset/go-ftw/blob/528e92cae88cd7c084e087856214f71f687cc4dc/runner/run.go#L371-L374), I believe it is reasonable not to look for triggered rules in this case.

M4tteoP commented on check/base.go:
```diff
@@ -116,6 +116,10 @@ func (c *FTWCheck) GetTriggeredRules() []uint {
 	if c.CloudMode() {
 		return nil
 	}
+	// When a test is expecting to trigger an error and it effectively does, markers are not set.
+	if len(c.log.StartMarker()) == 0 || len(c.log.EndMarker()) == 0 {
+		return nil
+	}
```
After https://github.com/coreruleset/go-ftw/pull/432, `GetTriggeredRules()` is now called not only when we are looking for IDs in the logs based on the test expectations but also at the end of the test to provide the list of triggered rules. When a test is expected to trigger an error, start and end markers might not populated. As a result, calling TriggeredRules() in such cases triggers a fatal error. Similar to how [we do not support looking for IDs when expecting an error](https://github.com/coreruleset/go-ftw/blob/528e92cae88cd7c084e087856214f71f687cc4dc/runner/run.go#L371-L374), I believe it is reasonable not to look for triggered rules in this case.

theseion replied:
Yes, I agree.

theseion commented on runner/run.go:
```diff
@@ -23,6 +23,13 @@ import (
 )
 
+const (
+	// start and end UUID suffixes are used to disambiguate start and end markers.
```
```suggestion
	// Start and end UUID suffixes are used to disambiguate start and end markers.
```

theseion commented on runner/run.go:
```diff
@@ -23,6 +23,13 @@ import (
 
+const (
+	// start and end UUID suffixes are used to disambiguate start and end markers.
+	// It permits to make the markers unique, while still maintaining one UUID per stage.
```
```suggestion
	// The suffixes make the markers unique, while still maintaining one UUID per stage.
```

theseion commented on waflog/read.go:
```diff
@@ -165,29 +170,38 @@ func (ll *FTWLogLines) getMarkedLines() [][]byte {
 		ll.markedLines = append(ll.markedLines, saneCopy)
 	}
+	if !startFound {
+		log.Debug().Msg("starter marker not found while getting marked lines")
```
```suggestion
		log.Debug().Msg("start marker not found while collecting marked lines")
```

theseion commented on waflog/read_test.go:
We don't have a test yet, that specifically reproduces the issue, right? Could you add one?

M4tteoP replied:
- With your idea of splitting the markers, the original issue ends up being an end marker not found by `CheckLogForMarker`, I slightly changed the already existing `TestReadCheckLogForMarkerNoMarkerAtEnd`.
- I added a new test (`TestReadCheckLogForMarkerWithMultipleMarkersAtEnd`) for the other potential issue that you [highlighted](https://github.com/coreruleset/go-ftw/pull/413#issuecomment-2530585028) (
